src/hotspot/share/ci/ciObject.cpp

```diff
@@ -168,6 +168,37 @@ jobject ciObject::constant_encoding() {
   return handle();
 }
 
+// ------------------------------------------------------------------
+// ciObject::check_constant_value_cache()
+//
+// Cache constant value lookups to ensure that consistent values are observed
+// during compilation because fields may be (re-)initialized concurrently.
+ciConstant ciObject::check_constant_value_cache(int off, BasicType bt) {
+  if (_constant_values != nullptr) {
+    for (int i = 0; i < _constant_values->length(); ++i) {
+      ConstantValue cached_val = _constant_values->at(i);
+      if (cached_val.off() == off) {
+        assert(cached_val.value().basic_type() == bt, "unexpected type");
+        return cached_val.value();
+      }
+    }
+  }
+  return ciConstant();
+}
+
+// ------------------------------------------------------------------
+// ciObject::add_to_constant_value_cache()
+//
+// Add a constant value to the cache.
+void ciObject::add_to_constant_value_cache(int off, ciConstant val) {
+  assert(val.is_valid(), "value must be valid");
+  if (_constant_values == nullptr) {
+    Arena* arena = CURRENT_ENV->arena();
+    _constant_values = new (arena) GrowableArray<ConstantValue>(arena, 0, 0, ConstantValue());
```

Comment:
Considering an element is added right away, it makes sense to increase initial capacity (to 1, at least).

Reply:
Makes sense. Done.